packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1741,6 +1748,96 @@ export class ContainerRuntime
 		this.removeAllListeners();
 	}
 
+	/**
+	 * Api to fetch the snapshot from the service for a loadingGroupIds.
+	 * @param loadingGroupIds - LoadingGroupId for which the snapshot is asked for.
+	 * @param pathParts - Parts of the path, which we want to extract from the snapshot tree.
+	 * @returns - snapshotTree and the sequence number of the snapshot.
+	 */
+	public async getSnapshotForLoadingGroupId(
+		loadingGroupIds: string[],
+		pathParts: string[],
+	): Promise<{ snapshotTree: ISnapshotTree; sequenceNumber: number }> {
+		const sortedLoadingGroupIds = loadingGroupIds.sort();
+		assert(this.storage.getSnapshot !== undefined, "getSnapshot api should be defined if used");
+		// Lookup up in the cache, if not present then make the network call as multiple datastores could
+		// be in same loading group. So, once we have fetched the snapshot for that loading group on
+		// any request, then cache that as same group could be requested in future too.
+		const snapshot = await this.snapshotCacheForLoadingGroupIds.addOrGet(
+			sortedLoadingGroupIds.join(),
+			async () => {
+				assert(
+					this.storage.getSnapshot !== undefined,
+					"getSnapshot api should be defined if used",
+				);
+				return this.storage.getSnapshot({
+					cacheSnapshot: false,
+					scenarioName: "snapshotForLoadingGroupId",
+					loadingGroupIds: sortedLoadingGroupIds,
+				});
+			},
+		);
+
+		// Find the snapshotTree inside the returned snapshot based on the path as given in the request.
+		const hasIsolatedChannels = rootHasIsolatedChannels(this.metadata);
+		const snapshotTreeForPath = this.getSnapshotTreeForPath(
+			snapshot.snapshotTree,
+			pathParts,
+			hasIsolatedChannels,
+		);
+		assert(snapshotTreeForPath !== undefined, "no snapshotTree for the path");
+		const snapshotSeqNumber = snapshot.sequenceNumber;
+		assert(snapshotSeqNumber !== undefined, "snapshotSeqNumber should be present");
+
+		// If the snapshot is ahead of the last seq number of the delta manager, then catch up before
+		// returning the snapshot.
+		if (snapshotSeqNumber > this.context.deltaManager.lastSequenceNumber) {
+			// If this is a summarizer client, which is trying to load a group and it finds that there is
+			// another snapshot from which the summarizer loaded and it is behind, then just give up as
+			// the summarizer state is not up to date.
+			if (this.isSummarizerClient) {
+				throw new Error(
+					"Summarizer client behind when loading snapshot with loadingGroupId",
+				);
+			}
+			// If the inbound deltas queue is paused, then unpause it for now as we need
+			// to catch up to the snapshot sequence number.
+			if (this.context.deltaManager.inbound.paused) {
```

Comment:
First, resume/pause is ref-counted. So doing this hear will for sure cause trouble as ref-counting will be off.

Depending on how it was paused, this might be not sufficient. For example, if Container.disconnect() was called (or equivalent - we never established connection in the first place), then this is not going to do anything.
While not ideal, we should stop poking at lowest layers (as much as we can) and expose and operate on much higher level (Container contracts).

It's up to us to decide what these contracts look like and what they do. Ideally there is one set of APIs (name-wise closer to pause() & unpause()), but under the covers they can drop and reestablish connection (for example, drop connection 15-60 seconds after pause() was called, unless unpause() was called).
Ideally even summarizer would use a contract like that (and single-commit summary to survive even if we disconnect from socket).

It's a bit shaky, but here are the key cases where we do pausing:
1. Summarizer: you got it covered
2. Schedule Manager - eventually it will be gone, but for now it's not causing trouble, as it operates on granularity of batches - so does summarizer (and thus snapshots). Thus, we can simply ignore it and system will be fine.
3. CoPilot (Bohemia repo) - we will need to move them off low-level APIs to something higher-level
4. Initial container load / Container.disconnect() flows - this (together with previous point) should be not-ref counted, and that's the layer this code needs to tap in in order to restore connection and move on.

IMHO, we have to hide low-level APIs.

@tyler-cai-microsoft - FYI

Reply:
So, for this change, I can remove the unpausing and just assume that we are ok?

Reply:
@vladsud I dug into what copilot was doing in bohemia, they seem to be only doing it in tests

Reply:
Great to hear that!

Reply:
100%, https://github.com/microsoft/FluidFramework/pull/19636/files, I put it in the changeset